src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs

```diff
@@ -380,10 +380,7 @@ private unsafe string AllocateStringForNameIndex(int index, out int dataOffset)
                     if (_ums.Position > _ums.Length - byteLen)
                         throw new BadImageFormatException(SR.Format(SR.BadImageFormat_ResourcesIndexTooLong, index));
 
-                    string? s = null;
-                    char* charPtr = (char*)_ums.PositionPointer;
-
-                    s = new string(charPtr, 0, byteLen / 2);
+                    string s = Encoding.Unicode.GetString(_ums.PositionPointer, byteLen);
```

Comment:
This is potential breaking change and performance regression. `Encoding.Unicode.GetString` does validation of the encoding. It will both break cases where there are binary blobs stored as strings in resources, and also it will make reading of the resources slower.

Reply:
I can certainly change this to just to a plain byte-swap on big-endian instead.  I was just confused because in this same routine a few lines below, we also do Encoding.Unicode.GetString -- the only difference is whether the input stream is a UnmanagedMemoryStream or not.   Shouldn't the two cases be otherwise semantically equivalent?

Reply:
There are probably other subtle observable differences betwen the `UnmanagedMemoryStream` and non-`UnmanagedMemoryStream` paths. I do not think it is worth trying to fix them.

(The `UnmanagedMemoryStream` path is taken in 99+% of ResourceReader use cases. If we were to fix the differences, it would be probably better to snap to `UnmanagedMemoryStream` behavior, not the other way around.)

Reply:
Patch updated to leave the LE path unchanges and just do byte swaps on BE.   I'm not sure the StringBuilder approach is really the most efficient way here -- if you have any better suggestion, please let me know.

Reply:
This would be better than StringBuilder:
```
char[] arr = new char[byteLen / 2];
for (int i = 0; i < arr.Length; i++)
{
    arr[i] = (char)BinaryPrimitives.ReverseEndianness((short)charPtr[i])
}
s = new string(arr);
```
Also, I think it would also fine to just call `Encoding.Unicode.GetString` on the BE path. It would not be bug-for-bug compatible with the LE path, but it is unlikely to ever bubble up as a problem for the BE platforms.

Reply:
Patch updated as per your suggestion, thanks!